### PR TITLE
[mono][aot] Deduplicate runtime invoke wrappers on iOS

### LINKED ISSUE #83973: [mono][aot] Further optimizations for dedup improvement
## Description

This issue tracks further optimizations for the introduced dedup improvement. 

## ~~Build without a container assembly~~

By introducing https://github.com/dotnet/runtime/pull/83511, container AOT module is loaded by using assembly name. Therefore, there is no need for additional AOT assembly in a bundle anymore.

**UPDATE:** With the current code, it would be a high effort to decouple AOT image and an associated assembly.

## Runtime invoke wrappers on iOS

Current implementation doesn't cover cases where both shared method and its inflated instances could be deduplicated. During the compilation, the shared method is emitted in corresponding AOT module, while inflated instances are emitted in a container AOT module. During the runtime, the shared method can't be retrieved as it was searched in the container AOT module.

Here is an example. During the compilation, the following list of inflated instances emitted in the dedup AOT module.
```
(wrapper runtime-invoke) object <Module>:runtime_invoke_byte__this___int (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_void__this___int_byte (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_int__this___int (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_void__this___int_int (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_int__this___int_int (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_void__this___int_int_int (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_uint16__this___int (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_void__this___int_uint16 (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_bool__this___int (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_void__this___int_byte (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_sbyte__this___int (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_void__this___int_sbyte (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_char__this___int (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_void__this___int_uint16 (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_int16__this___int (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_void__this___int_int16 (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_uint__this___int (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_void__this___int_uint (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_long__this___int (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_void__this___int_long (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_ulong__this___int (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_void__this___int_long (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_intptr__this___int (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_void__this___int_intptr (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_uintptr__this___int (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_void__this___int_intptr (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_single__this___int (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_void__this___int_single (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_double__this___int (object,intptr,intptr,intptr)
(wrapper runtime-invoke) object <Module>:runtime_invoke_void__this___int_double (object,intptr,intptr,intptr)
```
During the runtime, `(wrapper runtime-invoke) void object:runtime_invoke_dynamic (intptr,intptr,intptr,intptr)` was searched in the dedup AOT module as it is `RUNTIME_INVOKE_WRAPPER`, but not emitted in the dedup AOT module. It could be addressed by updating `mono_aot_get_method` method to handle such cases. Alternative approach could be adding a dedup flag in `MonoMethod` struct but it looks like inefficient approach.

## Investigate gshared methods emitted into the dedup AOT module

Dedup AOT module should contain extra methods and generic instances only. However, the following gshared methods are emitted in addition to its instances.

```
System_Numerics_Vector_As_TFrom_GSHAREDVT_TTo_GSHAREDVT_System_Numerics_Vector_1_TFrom_GSHAREDVT
System_Runtime_Intrinsics_Vector64_As_TFrom_GSHAREDVT_TTo_GSHAREDVT_System_Runtime_Intrinsics_Vector64_1_TFrom_GSHAREDVT
System_Runtime_Intrinsics_Vector128_As_TFrom_GSHAREDVT_TTo_GSHAREDVT_System_Runtime_Intrinsics_Vector128_1_TFrom_GSHAREDVT
System_Runtime_Intrinsics_Vector256_As_TFrom_GSHAREDVT_TTo_GSHAREDVT_System_Runtime_Intrinsics_Vector256_1_TFrom_GSHAREDVT
```

# Tasks
 - [x] ~~Simplify build process by excluding dedup assembly~~
 - [ ] Deduplicate runtime invoke wrappers on iOS
 - [x] Investigate gshared methods emitted into the dedup AOT module

### PR DESCRIPTION
**Pull request by kotlarmilos**
This PR deduplicates runtime invoke wrappers on iOS, which aligns WASM and iOS implementations. Approximately 30 runtime invoke wrappers have been deduplicated, resulting in minimal size savings. According to the HelloiOS measurements, the reduction amounts to about 2kb or 0.01%, which could be within the margin of statistical error.

Such change had an issue with the CI in https://github.com/dotnet/runtime/pull/84304, even though the tests passed locally. 

Fixes https://github.com/dotnet/runtime/issues/83973

### CODE REVIEW

File: src/mono/mono/mini/aot-compiler.c [hunk 1]
```diff
@@ -4331,6 +4331,12 @@ collect_dedup_method (MonoAotCompile *acfg, MonoMethod *method)
 static int
 add_method_full (MonoAotCompile *acfg, MonoMethod *method, gboolean extra, int depth)
 {
+	if (collect_dedup_method (acfg, method))
```
Comment (vargaz) [inline src/mono/mono/mini/aot-compiler.c:4334]: What is the purpose of this change ?

Comment (kotlarmilos) [reply src/mono/mono/mini/aot-compiler.c:4334]: Before this change, during the runtime, `(wrapper runtime-invoke) void object:runtime_invoke_dynamic (intptr,intptr,intptr,intptr)` was being searched in the dedup AOT module as `RUNTIME_INVOKE_WRAPPER`, but it was not emitted in the dedup AOT module.

Dynamic runtime invoke wrapper has been added to the list of methods which need to be deduplicated by invoking `add_method_full`. Moving `collect_dedup_method` to `add_method_full` method deduplicates `runtime_invoke_dynamic` that has been added as below.
https://github.com/dotnet/runtime/blob/aaa1de18a7f97774fd44a1728dafc0d86894800e/src/mono/mono/mini/aot-compiler.c#L4864-L4867

**Review by vargaz — APPROVED** [review]

**Comment by SamMonoRT** [comment]
> The `System.Runtime` tests are failing randomly at different points on the `ios-arm64` CI, but they are passing when run locally on a device and on the `tvos-arm64` CI. Considering that we are still in the early preview stage, I suggest two possible options:
> 
> * Close the PR and don't consider deduplication of the runtime invoke wrappers. This means that the iOS implementation will remain inconsistent with the WASM implementation.
> * Merge the PR and if the issue continues to occur on the CI, revert the PR.
> 
> /cc: @steveisok @SamMonoRT @vargaz

Are the random failures happening only as part of your PR, or have we seen those random failures outside your PR? If only in PR, I say try to add more logging and see if we can pinpoint any relevant fixes.

**Comment by kotlarmilos** [comment]
Test retries indicate that the failures occur randomly. Based on the test logs, it is observed that the failures do not follow any pattern and can happen randomly during the test suite execution, without being tied to any particular test.

**Comment by kotlarmilos** [comment]
@akoeplinger @steveisok I think you could provide valuable feedback as you have encountered the TCP failure in the past.

**Comment by kotlarmilos** [comment]
I've merged this PR. If we encounter more frequent failures following the merge, I will revert the change.

/cc: @akoeplinger @steveisok

**Comment by steveisok** [comment]
Agreed, I think this change does have an impact on CI and needs reverted.
